frontend/app/src/App.tsx

```diff
@@ -887,47 +926,35 @@ export class App extends PureComponent<Props, State> {
       const config = newSessionProto.config as Config
       const themeInput = newSessionProto.customTheme as CustomThemeConfig
 
-      const baseUriParts = this.getBaseUriParts()
-      if (baseUriParts) {
-        const { basePath } = baseUriParts
-
-        const prevPageNameInPath = extractPageNameFromPathName(
-          document.location.pathname,
-          basePath
-        )
-        const prevPageName =
-          prevPageNameInPath === "" ? mainPage.pageName : prevPageNameInPath
-        // It is important to compare `newPageName` with the previous one encoded in the URL
-        // to handle new session runs triggered by URL changes through the `onHistoryChange()` callback,
-        // e.g. the case where the user clicks the back button.
-        // See https://github.com/streamlit/streamlit/pull/6271#issuecomment-1465090690 for the discussion.
-        if (prevPageName !== newPageName) {
-          // If embed params need to be changed, make sure to change to other parts of the code that reference preserveEmbedQueryParams
-          const queryString = preserveEmbedQueryParams()
-          const qs = queryString ? `?${queryString}` : ""
-
-          const basePathPrefix = basePath ? `/${basePath}` : ""
-
-          const pagePath = viewingMainPage ? "" : newPageName
-          const pageUrl = `${basePathPrefix}/${pagePath}${qs}`
-
-          window.history.pushState({}, "", pageUrl)
-        }
-      }
-
+      // mainPage must be a string as we're guaranteed at this point that
+      // newSessionProto.appPages is nonempty and has a truthy pageName.
+      // Otherwise, we'd either have no main script or a nameless main script,
+      // neither of which can happen.
+      const mainPage = newSessionProto.appPages[0] as AppPage
+      // We're similarly guaranteed that newPageName will be found / truthy
+      // here.
+      const newPageName = newSessionProto.appPages.find(
+        p => p.pageScriptHash === newPageScriptHash
+      )?.pageName as string
+      const viewingMainPage = newPageScriptHash === mainPage.pageScriptHash
+
+      this.maybeUpdatePageUrl(mainPage.pageName, newPageName, viewingMainPage)
       this.processThemeInput(themeInput)
+      this.setState({
+        allowRunOnSave: config.allowRunOnSave,
+        hideTopBar: config.hideTopBar,
+        toolbarMode: config.toolbarMode,
+        latestRunTime: performance.now(),
+        // If we're here, the fragmentIdsThisRun variable is always the
+        // empty array.
+        fragmentIdsThisRun,
+      })
+
       this.setState(
         {
```

Comment:
Do these need to be set in two different calls to `this.setState`? I'm not sure if there's a meaningful difference in splitting up the two, and if there is there should probably be a comment explaining why it's done since it seems pretty unintuitive to me.

Reply:
I can add a commnet. Basically, the [next PR](https://github.com/streamlit/streamlit/commit/a00eed3c6c185e48d5421beea3e80be8a0dde208#diff-1b95d4e0dda1e5928f99c8303599bd811f140fdd40e028fcfca3077f802d6210L957) will essentially remove it complete.y.